luaui/Widgets/unit_factory_quota.lua

```diff
@@ -0,0 +1,253 @@
+end
+
+local function orderDequeue(unitID, buildDefID, count)
+	while count > 0 do
+		count = count - 100
+
+		GiveOrderToUnit(unitID, -buildDefID, {}, { "right", "ctrl", "shift" })
+	end
+end
+
+local function clearFactQ(factID)
+    
+	local queue = GetRealBuildQueue(unitID)
+
+	if queue ~= nil then
+		for _, buildPair in ipairs(queue) do
+			local buildUnitDefID, count = next(buildPair, nil)
+
+			orderDequeue(unitID, buildUnitDefID, count)
+		end
+	end
+end
+
+local function putInFrontOfFactQ(...)
+    local factID, cmdID, params, opts = ...
+    local commandq = GetFactoryCommands(factID, -1)
+    local altCmds = {}
+    local others = {}
+    for _, cmd in ipairs(commandq) do
+        if cmd.options.alt then
+            altCmds[#altCmds+1] = cmd
+        else
+            others[#others+1] = cmd
+        end
+    end
+
+    clearFactQ(factID)
+
+    for i = #altCmds, 1, -1 do -- do alt queue backwards
+        local cmd = altCmds[i]
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+    GiveOrderToUnit(...)
+    for _, cmd in ipairs(others) do
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+end
+
+local function insertToFactQ(...)
+    local factID, cmdID, params, opts = ...
+    local currCmd, targetID = Spring.GetUnitWorkerTask(factID)
+    local insertnormally = true
+    if targetID and Spring.GetUnitStates(factID)["repeat"] then
+        local _, _, _, _, buildProgress = Spring.GetUnitHealth(targetID)
+        if buildProgress < maxBuildProg and metalcosts[-currCmd] and (buildProgress * metalcosts[-currCmd]) < maxMetal then -- 7.5 % is the most that it is willing to cancel, and maximally 500 metal
+            insertnormally = false
+        end
+    end
+    if insertnormally then
+        GiveOrderToUnit(...)
+    else
+        putInFrontOfFactQ(...)
+    end
+end
+
+local function fillQuotas()
+    for factID, quota in pairs(quotas) do
+        if isFactoryUsable(factID) then
+            for udefid, num in pairs(quota) do
+                if num == 0 then
+                    quota[udefid] = nil
+                end
+            end
+            if table.length(quota)>0 then
+                local function isBetter(q1, q2, k1, k2)
+                    return (table.length((builtUnits[factID] or {})[k1] or {})/q1) < (table.length((builtUnits[factID] or {})[k2] or {})/q2)
+                end
+                local quotaNum, uDefID = findMin(quota, isBetter)
+                if quotaNum > table.length((builtUnits[factID] or {})[uDefID] or {}) then
+                    insertToFactQ(factID, -uDefID, {}, {"alt"})
+                end
+            end
+        end
+    end
+end
+
+function widget:GameFrame(n)
+    if n % 4 == 0 then -- improve perfomance
+        fillQuotas()
+    end
+end
+
+----- handle resetting queue mode to normal
+local function isOnQuotaBuildMode(unitID)
+	return GetUnitCmdDescs(unitID)[FindUnitCmdDesc(unitID, CMD_QUOTA_BUILD_TOGGLE)].params[1]+0 == 1
```

Comment:
This same logic appears in the build menu widgets, is there a single, common location this logic can be extracted to? Perhaps in the `WG['Quotas']` table.